### ISSUE TITLE
Detail of BinaryDenseNet or BinaryResNet18E

### ISSUE THREAD

**Issue by LaVieEnRoseSMZ** [hpi-xnor/BMXNet-v2#8]
Hi, recently I read your newly realeased paper  "Back to Simplicity: How to Train Accurate BNNs from Scratch?" It is a quite good paper and inspires me a lot. 

However, I am a little confused about the implementation in this paper. I am not familiar with the code structure of MXNet, Could you please write a more detailed readme or a tutorial or anything similar which could explain the code and the training details? 

Thanks a lot in advance~

**Comment by Jopyth** [comment]
Hi, I agree the framework definitely could benefit from more tutorials and documentation. However it would be best if these hit the "sweet spot" which cover what you (or anyone else who stumbles upon this) is interested in. In the title you name the details of the networks, but in your comment you mention training in general? Which resources have you used/found already? Did you manage to build the framework on your machine?

In the meantime, here are a few links, that might help you get started, depending on what you are looking for:

- The [Examples](https://github.com/hpi-xnor/BMXNet-v2-examples) contain code for building the models ([ResNetE](https://github.com/hpi-xnor/BMXNet-v2-examples/blob/master/binary_models/resnet_e.py), [BinaryDenseNet](https://github.com/hpi-xnor/BMXNet-v2-examples/blob/master/binary_models/densenet.py)) and training on ImageNet or Cifar ([image_classification.py](https://github.com/hpi-xnor/BMXNet-v2-examples/blob/master/image_classification.py))
- The training script has a lot of parameters, you could check out the pages we have already created in the [Wiki](https://github.com/hpi-xnor/BMXNet-v2-wiki) (e.g. regarding [Hyperparameters](https://github.com/hpi-xnor/BMXNet-v2-wiki/blob/master/hyperparameters.md)). There are also pages with options you need to provide to the training script to train/reproduce the results of a particular model. [This page](https://github.com/hpi-xnor/BMXNet-v2-wiki/blob/master/Binary-Network-Results.md) contains lots of experiments. Unfortunately, I have not yet updated this page to include the latest experiment results, but you can find the configuration details, e.g. for BinaryDenseNet (reduction), in the supplementary material already.

Please let me know if this already helps, or whether you would like additional information.

**Comment by LaVieEnRoseSMZ** [comment]
Thanks very much for your patience and detailed answer. I have spent days to read code and reproduce your result using pytorch. And I already get 56.3% Top-1 accuracy and it indicates your solid work. But as reported in your paper, the Best Accuracy for ResNetE is 58.1% and I haven't found this result in wiki or [hyperparameters](https://github.com/hpi-xnor/BMXNet-v2-wiki/blob/master/Binary-Network-Results.md). The best is 56.7% in this page.

![image](https://user-images.githubusercontent.com/17960617/62762196-24c43580-babb-11e9-8128-59dc6852a66b.png)

Hope you will find time to update latest work~Thanks in advance.

**Comment by LaVieEnRoseSMZ** [comment]
Thanks very much for your patience and detailed answer. I have spent days to reproduce your work using pytorch and I already gets 56.3% using your hyperparameter in wiki which indicated your solid work. And I have found your supplementary meterial in arxiv.com and found detailed log of BinaryResNetE of 58.1% and I will reproduce this results too. 

Thanks again for your detailed log and detailed supplementary material and it really helps us a lot~

**Comment by lgeiger** [comment]
We also started trying out BinaryDenseNets and the first results seam promising.

@LaVieEnRoseSMZ To which supplementary materials are you referring to? It doesn't seam like they are included in the arXiv version: https://arxiv.org/pdf/1906.08637.pdf

**Comment by LaVieEnRoseSMZ** [comment]
I am referring to the url in the comments of arxiv paper [supplementary material](https://owncloud.hpi.de/s/1jrAUnqRAfg0TXH)

**Comment by lgeiger** [comment]
Thanks. That's helpful 👍

**Comment by LaVieEnRoseSMZ** [comment]
I still have one more problem in reading the code of [Binarylayer](https://github.com/hpi-xnor/BMXNet-v2/blob/683fb59d35c4f5b044662ccd912e991edd5ac4a8/python/mxnet/gluon/nn/binary_layers.py)

I can not find the defination and implementation of "det_sign" which is used to quantize activation and weight. Can you please show me the url of this part of code?

Thanks a lot in advance~

**Comment by Jopyth** [comment]
As described in [Overview of changes](https://github.com/hpi-xnor/BMXNet-v2#overview-of-changes), you can find the parts of the code for `det_sign` in this [commit](https://github.com/hpi-xnor/BMXNet-v2/commit/ada4ea1d4418cfdd6cbc6d0159e1a716cb01cd85#diff-151ed3672c6bd71b358e13771357696d).

**Comment by LaVieEnRoseSMZ** [comment]
One more question about gradient_cancel layer, does it exactly the same as the supplementary material describes:
![image](https://user-images.githubusercontent.com/17960617/62770264-43332c80-bacd-11e9-8a0e-60b99b5d5d46.png)

Thanks a lot for your patient answering~

**Comment by Jopyth** [comment]
The combination of `gradient_cancelling` + `det_sign` (e.g. [QActivation] with `det_sign` as activation function (https://github.com/hpi-xnor/BMXNet-v2/blob/683fb59d35c4f5b044662ccd912e991edd5ac4a8/python/mxnet/gluon/nn/binary_layers.py#L153-L162)) does exactly what is described in the above part. You can check the implementation of gradient cancelling [here](https://github.com/hpi-xnor/BMXNet-v2/blob/master/src/operator/contrib/gradient_cancel-inl.h#L97-L112).

**Comment by LaVieEnRoseSMZ** [comment]
Hi~I am reproducing BinaryDenseNet in the paper. When I go through the code, I find three version of densenet called densenet, densenet_x and densenet_y, what is the exact version used in you final experiments? And I couldn't find DenseNet28 anywhere, would you mind show me the code?

Thanks a lot in advance~

**Comment by yanghaojin** [comment]
> Hi~I am reproducing BinaryDenseNet in the paper. When I go through the code, I find three version of densenet called densenet, densenet_x and densenet_y, what is the exact version used in you final experiments? And I couldn't find DenseNet28 anywhere, would you mind show me the code?
> 
> Thanks a lot in advance~

please check the supplementary material published with our paper: https://owncloud.hpi.de/s/1jrAUnqRAfg0TXH
you will find the detailed network configuration, network architecture, logs of training etc.

**Comment by lgeiger** [comment]
Thanks for sharing all the details and the supplementary materials. We were able to reproduce your experiments (which unfortunately isn't the case with many other papers) 👍 

@LaVieEnRoseSMZ If you are looking for an reimplementation of this paper using [Larq](http://larq.dev) (a Keras and TensorFlow based BNN library), you can also checkout the pretrained models and training code at https://larq.dev/models/ and https://github.com/larq/zoo.
